python/ray/data/_internal/execution/operators/actor_pool_submitter.py

```diff
@@ -99,40 +254,179 @@ def submit(
         yield from _map_task(fn, *blocks)
 
 
-class ActorPool:
+@dataclass
+class AutoscalingConfig:
+    """Configuration for an autoscaling actor pool."""
+
+    # Minimum number of workers in the actor pool.
+    min_workers: int
+    # Maximum number of workers in the actor pool.
+    max_workers: int
+    # Maximum number of tasks that can be in flight for a single worker.
+    # TODO(Clark): Have this informed by the prefetch_batches configuration, once async
+    # prefetching has been ported to this new actor pool.
+    max_tasks_in_flight: int = 2
+    # Minimum ratio of ready workers to the total number of workers. If the pool is
+    # above this ratio, it will be allowed to be scaled up.
+    # NOTE: This is set to a lower value than the legacy actor compute strategy since we
+    # need to more aggressively scale-up with an event-based autoscaler rather than a
+    # 10ms control loop autoscaler.
+    ready_to_total_workers_ratio: float = 0.5
+    # Maximum ratio of idle workers to the total number of workers. If the pool goes
+    # above this ratio, the pool will be scaled down.
+    idle_to_total_workers_ratio: float = 0.5
+
+    def __post_init__(self):
+        if self.min_workers < 1:
+            raise ValueError("min_workers must be >= 1, got: ", self.min_workers)
+        if self.max_workers is not None and self.min_workers > self.max_workers:
+            raise ValueError(
+                "min_workers must be <= max_workers, got: ",
+                self.min_workers,
+                self.max_workers,
+            )
+        if self.max_tasks_in_flight < 1:
+            raise ValueError(
+                "max_tasks_in_flight must be >= 1, got: ",
+                self.max_tasks_in_flight,
+            )
+
+
+class AutoscalingPolicy:
+    """Autoscaling policy for an actor pool, determining when the pool should be scaled
+    up and when it should be scaled down.
+    """
+
+    def __init__(self, autoscaling_config: "AutoscalingConfig"):
+        self._config = autoscaling_config
+
+    @property
+    def min_workers(self) -> int:
+        """The minimum number of actors that must be in the actor pool."""
+        return self._config.min_workers
+
+    @property
+    def max_workers(self) -> int:
+        """The maximum number of actors that can be added to the actor pool."""
+        return self._config.max_workers
+
+    def should_scale_up(self, num_total_workers: int, num_running_workers: int) -> bool:
+        """Whether the actor pool should scale up by adding a new actor.
+
+        Args:
+            num_total_workers: Total number of workers in actor pool.
+            num_running_workers: Number of currently running workers in actor pool.
```

Comment:
If an actor is running less than its max number of tasks, is it included here?

Reply:
Yep, `num_running_workers` should include both idle and active workers, i.e. any worker in the pool that is no longer pending.